Saturation/SaturationAlgorithm.cpp

```diff
@@ -959,13 +1011,37 @@ bool SaturationAlgorithm::forwardSimplify(Clause* cl)
         if (replacement) {
           addNewClause(replacement);
         }
-        onClauseReduction(cl, replacement, premises);
+        onClauseReduction(cl, pvi(getSingletonIterator(replacement)), premises);
 
         return false;
       }
     }
   }
 
+  static ClauseStack repStack;
+
+  repStack.reset();
+  SimplList::Iterator sit(_simplifiers);
+
+  while (sit.hasNext()) {
+    SimplificationEngine* se=sit.next();
+
+    {
+      ClauseIterator results = se->perform(cl);
+ 
+      if (results.hasNext()) {
+        while(results.hasNext()){
+          Clause* simpedCl = results.next();
+          ASS(simpedCl != cl);
+          repStack.push(simpedCl);
+          addNewClause(simpedCl);
+        }
+        onClauseReduction(cl, pvi(ClauseStack::Iterator(repStack)), 0);
```

Comment:
You put the replacements on a stack, turn that into an iterator, then pass it to onClauseReduction and then turn the iterator back into a stack. 

Did you consider passing `const& ClauseStack` directly instead?

Reply:
Or simply (Clause*)[] and a count?

Reply:
That is quite strange!

I think the reason for using an iterator is that it is easier to set up single element iterators than single element stacks, but I will attempt to change this.

Reply:
Note that a single element (Clause*)[] is just the address of the pointer holding the clause in the current context. 

Though this can only be used, of course, if the caller lives at all times when the reference is needed. But I think this is the case here!